Include/internal/pycore_frame.h

```diff
@@ -20,13 +20,13 @@ enum _framestate {
 typedef signed char PyFrameState;
 
 typedef struct _interpreter_frame {
-    PyObject *f_globals;
-    PyObject *f_builtins;
-    PyObject *f_locals;
-    PyCodeObject *f_code;
-    PyFrameObject *frame_obj;
-    /* Borrowed reference to a generator, or NULL */
-    PyObject *generator;
+    PyFunctionObject *f_func; /* Strong reference */
+    PyObject *f_globals; /* Borrowed reference */
+    PyObject *f_builtins; /* Borrowed reference */
```

Comment:
So, we avoid extra refcount operations for these, since `f_func` is guaranteed to hold a ref, right?  Cool!

The only problem I see is that the function's `func_globals` (or `func_builtins`) could get swapped out in another thread while this frame is executing.  Then we'd lose that borrowed reference and potentially segfault.  It's an unlikely case, but still a possibility.

Reply:
Having these on `struct _interpreter_frame` is helpful because we avoid extra pointer indirection, right?

However, that only matters in cases where these are used relatively frequently (so the cost of the indirection adds up), i.e. in the eval loop (via `GLOBALS()`).  If we do not expect `frame->f_globals` to change during the eval loop then could we move both to local variables before starting the loop (and drop them from `struct _interpreter_frame` altogether)?

Note that then we would have a different possible race where the function's `func_globals` could get swapped out while the frame is running, leading to weirdness.

Reply:
Also, why don't we similarly add `f_closure` to the frame?  Is it because we want to avoid increasing the size of `struct _interpreter_frame` (and we don't actually use the closure that much, so the cost of the indirection is bearable)?

Reply:
> Also, why don't we similarly add `f_closure` to the frame? Is it because we want to avoid increasing the size of `struct _interpreter_frame` (and we don't actually use the closure that much, so the cost of the indirection is bearable)?

We might want to avoid creating a closure object at all. We could put the cells at the end of the function object, rather than in a separate tuple. but that's future work.

Reply:
Looks like you skipped over the two other comments here.

Reply:
The frame has a strong reference to the function, which has a strong reference to both the globals and builtins.
So borrowed references to the globals and builtins are OK.

> If we do not expect frame->f_globals to change during the eval loop then could we move both to local variables before starting the loop (and drop them from struct _interpreter_frame altogether)?

And where would you store those references across calls?

Reply:
If we keep `f_func` on the frame then that's where we'd get them.